NorthstarDedicatedTest/sv_rcon.cpp

```diff
@@ -0,0 +1,569 @@
+// Purpose: authenticate new connections
+// Input  : *cl_request -
+//			*pData -
+//-----------------------------------------------------------------------------
+void CRConServer::Authenticate(const cl_rcon::request& cl_request, CConnectedNetConsoleData* pData)
+{
+	if (pData->m_bAuthorized)
+	{
+		return;
+	}
+	else // Authorize.
+	{
+		if (this->Comparator(cl_request.requestbuf()))
+		{
+			pData->m_bAuthorized = true;
+			m_pSocket->CloseListenSocket();
+
+			this->CloseNonAuthConnection();
+			this->Send(pData->m_hSocket, this->Serialize(s_pszAuthMessage, "", sv_rcon::response_t::SERVERDATA_RESPONSE_AUTH));
+		}
+		else // Bad password.
+		{
+			CNetAdr2 netAdr2 = m_pSocket->GetAcceptedSocketAddress(m_nConnIndex);
+			if (CVar_sv_rcon_debug->GetBool())
+			{
+				spdlog::info("Bad RCON password attempt from '{:s}'", netAdr2.GetIPAndPort());
+			}
+
+			this->Send(pData->m_hSocket, this->Serialize(s_pszWrongPwMessage, "", sv_rcon::response_t::SERVERDATA_RESPONSE_AUTH));
+
+			pData->m_bAuthorized = false;
+			pData->m_bValidated = false;
+			pData->m_nFailedAttempts++;
+		}
+	}
+}
+
+//-----------------------------------------------------------------------------
+// Purpose: sha256 hashed password comparison
+// Input  : *svCompare -
+// Output : true if matches, false otherwise
+//-----------------------------------------------------------------------------
+bool CRConServer::Comparator(std::string svPassword) const
+{
+	svPassword = sha256(svPassword);
+	if (CVar_sv_rcon_debug->GetBool())
+	{
+		spdlog::info("+---------------------------------------------------------------------------+");
+		spdlog::info("] Server: '{:s}'[", m_svPasswordHash);
+		spdlog::info("] Client: '{:s}'[", svPassword);
+		spdlog::info("+---------------------------------------------------------------------------+");
+	}
+	if (memcmp(svPassword.c_str(), m_svPasswordHash.c_str(), SHA256::DIGEST_SIZE) == 0)
+	{
+		return true;
+	}
+	return false;
+}
+
+//-----------------------------------------------------------------------------
+// Purpose: parses input response buffer using length-prefix framing
+// Input  : *pRecvBuf -
+//			nRecvLen -
+//			*pData -
+//-----------------------------------------------------------------------------
+void CRConServer::ProcessBuffer(const char* pRecvBuf, int nRecvLen, CConnectedNetConsoleData* pData)
+{
+	while (nRecvLen > 0)
+	{
+		if (pData->m_nPayloadLen)
+		{
+			if (pData->m_nPayloadRead < pData->m_nPayloadLen)
+			{
+				pData->m_RecvBuffer[pData->m_nPayloadRead++] = *pRecvBuf;
+
+				pRecvBuf++;
+				nRecvLen--;
+			}
+			if (pData->m_nPayloadRead == pData->m_nPayloadLen)
+			{
+				this->ProcessMessage(
+					this->Deserialize(std::string(reinterpret_cast<char*>(pData->m_RecvBuffer.data()), pData->m_nPayloadLen)));
+
+				pData->m_nPayloadLen = 0;
+				pData->m_nPayloadRead = 0;
+			}
+		}
+		else if (pData->m_nPayloadRead < sizeof(int)) // Read size field.
+		{
+			pData->m_RecvBuffer[pData->m_nPayloadRead++] = *pRecvBuf;
```

Comment:
> Expression: vector subscript out of range

Because `m_RecvBuffer` has 4 bytes reserved, but it's length is still 0. I was able to just change the `CConnectedNetConsoleData` constructor to initialize the vector to 4 elements instead of reserving space, but I'm not sure if that's the best way with how you're using the buffer.

Reply:
Thanks, small oversight. Fixed as of commit 24f561f8bb656bd7a242d77c40651318e89ad2f8.